extensions/dicom-sr/src/OHIFCornerstoneSRViewport.js

```diff
@@ -229,22 +251,158 @@ function OHIFCornerstoneSRViewport({
+      displaySet.StudyInstanceUID,
+      displaySet.SeriesInstanceUID,
+      displaySet.SOPInstanceUID
+    );
+
+    const { MeasurementReport } = adapters.Cornerstone;
+
+    const sopInstanceUIDToImageId = {};
+
+    displaySet.measurements.forEach(measurement => {
+      const { ReferencedSOPInstanceUID, imageId } = measurement;
+      if (!sopInstanceUIDToImageId[ReferencedSOPInstanceUID]) {
+        sopInstanceUIDToImageId[ReferencedSOPInstanceUID] = imageId;
+      }
+    });
+
+    // Use dcmjs to generate toolState.
+    const storedMeasurementByToolType = MeasurementReport.generateToolState(
+      instance
+    );
+
+    // Filter what is found by DICOM SR to measurements we support.
+    const mappingDefinitions = mappings.map(m => m.definition);
+    const hydratableMeasurementsInSR = {};
+
+    Object.keys(storedMeasurementByToolType).forEach(key => {
+      if (mappingDefinitions.includes(key)) {
+        hydratableMeasurementsInSR[key] = storedMeasurementByToolType[key];
+      }
+    });
+
+    if (
+      extensionManager.registeredExtensionIds.includes(
+        MEASUREMENT_TRACKING_EXTENSION_ID
+      )
+    ) {
+      // Set the series touched as tracked.
+      const imageIds = [];
+
+      Object.keys(hydratableMeasurementsInSR).forEach(toolType => {
+        const toolDataForToolType = hydratableMeasurementsInSR[toolType];
+
+        toolDataForToolType.forEach(data => {
+          // Add the measurement to toolState
+          const imageId = sopInstanceUIDToImageId[data.sopInstanceUid];
+
+          if (!imageIds.includes(imageId)) {
+            imageIds.push(imageId);
+          }
+        });
+      });
+
+      let targetStudyInstanceUID;
+      const SeriesInstanceUIDs = [];
+
+      for (let i = 0; i < imageIds.length; i++) {
+        const imageId = imageIds[0];
+        const {
+          SeriesInstanceUID,
+          StudyInstanceUID,
+        } = cornerstone.metaData.get('instance', imageId);
+
+        if (!SeriesInstanceUIDs.includes(SeriesInstanceUID)) {
+          SeriesInstanceUIDs.push(SeriesInstanceUID);
+        }
+
+        if (!targetStudyInstanceUID) {
+          targetStudyInstanceUID = StudyInstanceUID;
+        } else if (targetStudyInstanceUID !== StudyInstanceUID) {
+          console.warn(
+            'NO SUPPORT FOR SRs THAT HAVE MEASUREMENTS FROM MULTIPLE STUDIES.'
+          );
+        }
+      }
+
+      sendTrackedMeasurementsEvent('SET_TRACKED_SERIES', {
+        StudyInstanceUID: targetStudyInstanceUID,
+        SeriesInstanceUIDs,
+      });
+    }
+
+    Object.keys(hydratableMeasurementsInSR).forEach(toolType => {
+      const toolDataForToolType = hydratableMeasurementsInSR[toolType];
+
+      toolDataForToolType.forEach(data => {
+        // Add the measurement to toolState
+        const imageId = sopInstanceUIDToImageId[data.sopInstanceUid];
+
+        data.id = guid();
+
+        _addToolDataToCornerstoneTools(data, toolType, imageId);
```

Comment:
Is the flow here... 

DCMJS --> map to csTools shape w/ adapter 
mapped --> `cornerstoneTools` 
mapped --> addRaw to MeasurementService

This may be more challenging, but I was expecting:


DCMJS --> map to csTools shape w/ adapter --> add raw to `MeasurementService` 
SUBSCRIBED --> CornerstoneTools updates

Reply:
Yeah that's the current flow. We can discuss later, but since `dcmjs` already provides toolState in the exact form needed, this was a lot simpler. The latter requires an additional mapping.

Also I don't know if its particularly bad to be able to add state to a particular framework, and then tell the MeasurementService to track it, which is what we are doing here. Its not different from drawing cornerstoneTools annotations and then tracking them, just the source is external.